server/utils/auth.py

```diff
@@ -0,0 +1,117 @@
+from os import abort
+from flask import redirect, request, session, url_for, render_template
+from flask_login import LoginManager, login_user, logout_user, login_required
+from flask_oauthlib.client import OAuth
+from oauth2client.contrib.flask_util import UserOAuth2
+
+
+from server import app
+from server.models import db, User
+import server.utils.canvas as canvas_client
+from server.utils.stub import get_dev_user_oauth_resp
+from server.form import DevLoginForm
+
+login_manager = LoginManager(app=app)
+
+oauth = OAuth()
+
+server_url = app.config.get('CANVAS_SERVER_URL')
+consumer_key = app.config.get('CANVAS_CLIENT_ID')
+consumer_secret = app.config.get('CANVAS_CLIENT_SECRET')
+
+canvas_oauth = oauth.remote_app(
+    'seating',
+    consumer_key=consumer_key,
+    consumer_secret=consumer_secret,
+    base_url=server_url,
+    request_token_url=None,
+    access_token_method='POST',
+    access_token_url=server_url + 'login/oauth2/token',
+    authorize_url=server_url + 'login/oauth2/auth',
+)
+
+
+@canvas_oauth.tokengetter
+def get_access_token(token=None):
+    return session.get('access_token')
+
+
+google_oauth = UserOAuth2(app)
+
+
+@login_manager.user_loader
+def load_user(user_id):
+    return User.query.get(user_id)
+
+
+@login_manager.unauthorized_handler
+def unauthorized():
+    session['after_login'] = request.url
```

Comment:
Where does request.url come from? This looks like a potential open redirect vulnerability

Reply:
I am actually not sure. I will need to look into this.